cairo/src/runner.rs

```diff
@@ -0,0 +1,472 @@
+            // execute current step and increase time counter
+            step.execute();
+            n += 1;
+            match step.next {
+                None => end = true, // if find no next pointers, end
+                _ => {
+                    // if there are next pointers
+                    end = false;
+                    // update next value of pointers
+                    next = step.next.unwrap();
+                    if curr.ap <= next.pc {
+                        // if reading from unallocated memory, end
+                        end = true;
+                    }
+                }
+            }
+        }
+        self.steps = F::from(n);
+        self.fin = CairoPointers::new(curr.pc, curr.ap, curr.fp);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::helper::CairoFieldHelpers;
+    use mina_curves::pasta::fp::Fp as F;
+
+    #[test]
+    fn test_cairo_step() {
+        // This tests that CairoStep works for a 2 word instruction
+        //    tempvar x = 10;
+        let instrs = vec![
+            F::from(0x480680017fff8000u64),
+            F::from(10u64),
+            F::from(0x208b7fff7fff7ffeu64),
+        ];
+        let mut mem = CairoMemory::new(instrs);
+        // Need to know how to find out
+        // Is it final ap and/or final fp? Will write to starkware guys to learn about this
+        mem.write(F::from(4u32), F::from(7u32));
+        mem.write(F::from(5u32), F::from(7u32));
+        let ptrs = CairoPointers::new(F::from(1u32), F::from(6u32), F::from(6u32));
+        let mut step = CairoStep::new(&mut mem, ptrs);
+
+        step.execute();
+        assert_eq!(step.next.unwrap().pc, F::from(3u32));
+        assert_eq!(step.next.unwrap().ap, F::from(7u32));
+        assert_eq!(step.next.unwrap().fp, F::from(6u32));
+
+        println!("{}", step.mem);
+    }
+
+    #[test]
+    fn test_cairo_program() {
+        let instrs = vec![
+            F::from(0x480680017fff8000u64),
+            F::from(10u64),
+            F::from(0x208b7fff7fff7ffeu64),
+        ];
+        let mut mem = CairoMemory::<F>::new(instrs);
+        // Need to know how to find out
+        // Is it final ap and/or final fp? Will write to starkware guys to learn about this
+        mem.write(F::from(4u32), F::from(7u32)); //beginning of output
+        mem.write(F::from(5u32), F::from(7u32)); //end of output
+        let prog = CairoProgram::new(&mut mem, 1, 6);
+        println!("{}", prog.mem);
+    }
+
+    #[test]
+    fn test_cairo_output() {
+        // This is a test for a longer program, involving builtins, imports and outputs
+        /*
+        %builtins output
+        from starkware.cairo.common.serialize import serialize_word
+        func main{output_ptr : felt*}():
+            tempvar x = 10
+            tempvar y = x + x
+            tempvar z = y * y + x
+            serialize_word(x)
+            serialize_word(y)
+            serialize_word(z)
+            return ()
+        end
```

Comment:
if possible, include a link to the UI from where I can easily recompile this code and obtain the instructions below (for example if I want to modify the test)

Reply:
Oh sure I can :)